runtime/src/main/java/org/corfudb/runtime/collections/Query.java

```diff
@@ -0,0 +1,449 @@
+    }
+
+    /**
+     * Execute a join of 2 tables.
+     *
+     * @param tableName1     Table name 1.
+     * @param tableName2     Table name 2.
+     * @param query1         Predicate to filter entries in table 1.
+     * @param query2         Predicate to filter entries in table 2.
+     * @param joinPredicate  Predicate to filter entries during the join.
+     * @param joinFunction   Function to merge entries.
+     * @param joinProjection Project the merged entries.
+     * @param <Val1>         Type of Value in table 1.
+     * @param <Val2>         Type of Value in table 2.
+     * @param <T>            Type of resultant value after merging type V and type W.
+     * @param <U>            Type of value projected from T.
+     * @return Result of query.
+     */
+    @Nonnull
+    public <Val1 extends Message, Val2 extends Message, Meta1 extends Message, Meta2 extends Message, T, U>
+    QueryResult<U> executeJoinQuery(
+            @Nonnull final String tableName1,
+            @Nonnull final String tableName2,
+            @Nonnull final Predicate<CorfuRecord<Val1, Meta1>> query1,
+            @Nonnull final Predicate<CorfuRecord<Val2, Meta2>> query2,
+            @Nonnull final BiPredicate<Val1, Val2> joinPredicate,
+            @Nonnull final BiFunction<Val1, Val2, T> joinFunction,
+            final Function<T, U> joinProjection) {
+        return executeJoinQuery(
+                tableName1,
+                tableName2,
+                query1,
+                query2,
+                DEFAULT_OPTIONS,
+                DEFAULT_OPTIONS,
+                joinPredicate,
+                joinFunction,
+                joinProjection);
+    }
+
+    /**
+     * Execute a join of 2 tables.
+     *
+     * @param tableName1     Table name 1.
+     * @param tableName2     Table name 2.
+     * @param query1         Predicate to filter entries in table 1.
+     * @param query2         Predicate to filter entries in table 2.
+     * @param queryOptions1  Query options to transform table 1 filtered values.
+     * @param queryOptions2  Query options to transform table 2 filtered values.
+     * @param joinPredicate  Predicate to filter entries during the join.
+     * @param joinFunction   Function to merge entries.
+     * @param joinProjection Project the merged entries.
+     * @param <Val1>         Type of Value in table 1.
+     * @param <Val2>         Type of Value in table 2.
+     * @param <R>            Type of projected values from table 1 from type V.
+     * @param <S>            Type of projected values from table 2 from type W.
+     * @param <T>            Type of resultant value after merging type R and type S.
+     * @param <U>            Type of value projected from T.
+     * @return Result of query.
+     */
+    @Nonnull
+    public <Val1 extends Message, Val2 extends Message, Meta1 extends Message, Meta2 extends Message, R, S, T, U>
+    QueryResult<U> executeJoinQuery(
```

Comment:
We can just call it join to be consistent with other APIs.

Reply:
want to keep same name as what the clients are familiar with if that's ok?